### ISSUE TITLE
different input structure of cmd file out, vdf

### ISSUE THREAD

**Issue by hyunjimoon** [hyunjimoon/VaccineMisinf#5]
why not out files are explicitly written in cmd file? danger of being overwritten!!
```
SPECIAL>NOINTERACTION
SPECIAL>LOADMODEL|CovidUSA-Econ-V79.mdl
SIMULATE>payoff|PayoffFlowV77-Policy.vpd
SIMULATE>sensitivity|
SIMULATE>optparm|COVID-V79-All.voc
SIMULATE>savelist|
SIMULATE>senssavelist|
SIMULATE>DATA|"CovidModelInputs - ConstantDataStates.vdf,CovidModelInputs - CRWStates.vdf,CovidModelInputs - DeathDataStates.vdf,CovidModelInputs - FlowDataStates.vdf,CovidModelInputs - FormattedDataStates.vdf,CovidModelInputs - TestDataStates.vdf"
SIMULATE>DATA|"V
SIMULATE>SETVAL|VaxInc=0
SIMULATE>SETVAL|Sensitivity to Weather=2.85234

SIMULATE>RUNNAME|Covidparam_Sensitivity
SIMULATE>REPORT|1
MENU>RUN|o
MENU>VDF2CSV|!|!|../SAVEFILE.txt||||

```

**Comment by hyunjimoon** [comment]
<img width="392" alt="image" src="https://user-images.githubusercontent.com/30194633/211165871-60bbf8c1-ff3c-48fd-a4d3-36fdb346c781.png">

need to press simulate for .out and .vdfx

**Comment by hyunjimoon** [comment]
second outfile comes from interface within vensim, and the line12 .vdf file is from cmdscript 
```
10:V79-OptimTest.out,VAMA\Calib_econlearning_Test_only1.out
11:COVID-V79-All.voc
12:PayoffFlowV77-Policy.vpd
13:CovidModelInputs - ConstantDataStates.vdf,CovidModelInputs - CRWStates.vdf,CovidModelInputs - DeathDataStates.vdf,CovidModelInputs - FlowDataStates.vdf,CovidModelInputs - FormattedDataStates.vdf,CovidModelInputs - TestDataStates.vdf,GDP_Vensim.vdfx
```

**Comment by hyunjimoon** [comment]
Summary of lessons:
- close every csv file while running VST script; excel doesn't let it overwrite
- mind the different input structure of .out, .vdf (.out is controlled with Vensim GUI, .vdf is with VST generated command script)
- inconsistency of econ-only (E2) and epidemic+econ (E12) model cause failure of attaching .out file. E12 has two `.out` (E1 calib, E2 calib), seven `.vdf` (6 driving E1, 1 driving E2) file in total.

**Comment by hyunjimoon** [comment]
@tseyanglim, could you confirm whether 
1.`.out` file can be supplied through command file?
```
{
	"basename": "Covid",
	"simcontrol": {
		"model": "CovidUSA-Econ-V79-Updated.mdl",
		"data": [
			"CovidModelInputs - ConstantDataStates.vdf",
			"CovidModelInputs - CRWStates.vdf",
			"CovidModelInputs - DeathDataStates.vdf",
			"CovidModelInputs - FlowDataStates.vdf",
			"CovidModelInputs - FormattedDataStates.vdf",
			"CovidModelInputs - TestDataStates.vdf",
			"GDP_Vensim.vdf"
			], 
		"payoff": "PayoffFlowV77-Policy.vpd",
		"sensitivity": "", 
		"optparm": "COVID-V79-All.voc",
		"changes": [], 
		"savelist": "", 
		"senssavelist": ""
		},
	"venginepath": "C:/Program Files (x86)/Vengine2/Vensim - vengine.exe",
	"runcmd": "",
	"savecmd": "VDF2CSV|!|!|SAVEFILE.txt|||",
	"timelimit": 600,
	"mccores": 0
	}
```

2. two outfiles (calibrated results), one from epidemic-only model and second from epidemic+econ model

**Comment by tseyanglim** [comment]
> @tseyanglim, could you confirm whether 1.`.out` file can be supplied through command file?
> 
> ```
> {
> 	"basename": "Covid",
> 	"simcontrol": {
> 		"model": "CovidUSA-Econ-V79-Updated.mdl",
> 		"data": [
> 			"CovidModelInputs - ConstantDataStates.vdf",
> 			"CovidModelInputs - CRWStates.vdf",
> 			"CovidModelInputs - DeathDataStates.vdf",
> 			"CovidModelInputs - FlowDataStates.vdf",
> 			"CovidModelInputs - FormattedDataStates.vdf",
> 			"CovidModelInputs - TestDataStates.vdf",
> 			"GDP_Vensim.vdf"
> 			], 
> 		"payoff": "PayoffFlowV77-Policy.vpd",
> 		"sensitivity": "", 
> 		"optparm": "COVID-V79-All.voc",
> 		"changes": [], 
> 		"savelist": "", 
> 		"senssavelist": ""
> 		},
> 	"venginepath": "C:/Program Files (x86)/Vengine2/Vensim - vengine.exe",
> 	"runcmd": "",
> 	"savecmd": "VDF2CSV|!|!|SAVEFILE.txt|||",
> 	"timelimit": 600,
> 	"mccores": 0
> 	}
> ```
> 
> 2. two outfiles (calibrated results), one from epidemic-only model and second from epidemic+econ model

1. For reading in `.out` files as input into a model run, you can supply as many as you want, via the `changes` list under `simcontrol`.
2. If you mean writing output to different `.out` files, that can't be done straightforwardly using just a single control file (or single version of a control file). One set of controls = one `Script` object = one model run = one set of model outputs.

**Comment by hyunjimoon** [comment]
>If you mean writing output to different .out 
I meant supply two .out files (which you answered in 1 that "changes":["epidemic_calib.out", "epid-econ_calib.out" ]), but 2 is good to know.

3. does vst read info from .mdl file and rewrite control file (`.cmd) everytime?
This was what I was expecting but below is my .mdl file, but when I ran VST with this, it didn't add .out info to control file. The first below is `.mdl`, second is `control.txt` generated.
<img width="674" alt="image" src="https://user-images.githubusercontent.com/30194633/212130649-72641bb8-90a1-4812-a5e3-1339ac862ba0.png">

**Comment by tseyanglim** [comment]
>I meant supplying two .out files (which you answered in 1 that "changes":["epidemic_calib.out", "epid-econ_calib.out" ]), but 2 is good to know.

Yep, answer #1 is what you need then, just put them in `changes` (alongside any `.cin` or other changes files you want).

>does vst read info from .mdl file about constant changes and data' and rewrite control file (.cmd) everytime?
This was what I was expecting but below is my .mdl file, but when I ran VST with this, it didn't add .out info to control file. The first below is .mdl, second is control.txt generated.

If I'm understanding the question correctly, the answer is no. The `.cmd` generated from the controlfile (and any modifications to the `Script` object) will _override_ whatever sim control settings are in the `.mdl` file. So it ignores any changes, data, etc. files that are loaded in the `.mdl` file.

You can think of the `simcontrol` portion of the controlfile / `Script` object as essentially doing the same thing as the sim control dialogue boxes in Vensim. So any setting of sensitivity control, optimization control, data, etc. will have to be input through the `simcontrol`.

Note that this is both a feature of `.cmd` files themselves, and an intentional choice - we want everything that's going into the `.cmd` file to be transparent in the Python script and controlfile, and if it read info from the `.mdl` file, that could easily change without the user being aware of it and create weird outputs that would be hard to trace.

**Comment by hyunjimoon** [comment]
> read info from the .mdl file,
@tseyanglim do you have any code that I can borrow that reads parameter value directly from `.mdl` file? I did manually inside vensim > create .tab file > read from it, but I'm sure VST would have this function.

**Comment by tseyanglim** [comment]
Yep, `get_value()` is your best bet for general purpose reading of a value from an `.mdl` file.

It should work with most Vensim-associated file formats, including `.mdl`, `.out`, `.cin`, `.voc`; anything where the relevant values are contained as `var = val` - it should be fine for different file formats as long as that condition is met.

Be careful with variable names; I believe it _should_ work with numbers, special characters, etc. but I haven't tested it. Also for subscripted variables it may be finicky - if it's something like `Variable[subscript] = X, Y, Z` it will fail (or only read the first value); if it's a single element or a fixed value for the whole subscript (either `Variable[element] = X` or `Variable[subscript] = X`) it should work, though in the first case you'll have to feed in the element-specific variable name.
